### PR TITLE
Add flamegraph profiling using flame and flamer

### PR DESCRIPTION
**Pull request by coolreader18**
This is just an initial change to get the ball rolling for flame profiling, we can add more scopes later.

### CODE REVIEW

**Comment by coolreader18** [comment]
There's not a ton of information here that would be helpful to optimization (yet), but it's really interesting to see!

image

**Comment by windelbouwman** [comment]
This is pretty cool! Is it mandatory to add the flame-it attribute to each function you wish to track? Can we just follow all functions?

File: vm/src/frame.rs [hunk 1]
```diff
@@ -268,7 +268,15 @@ impl Frame {
         }
     }
 
+    // #[cfg_attr(feature = "flame-it", flame("Frame"))]
     pub fn run(&self, vm: &VirtualMachine) -> Result<ExecutionResult, PyObjectRef> {
+        flame_guard!(format!("Frame::run({})", self.code.obj_name));
+        // flame doesn't include notes in the html graph :(
+        flame_note!(
+            flame::StrCow::from("CodeObj name"),
+            self.code.obj_name.clone().into()
```
Comment (windelbouwman) [inline vm/src/frame.rs]: Nice! Now we can follow python objects as well!

File: vm/src/frame.rs [hunk 2]
```diff
@@ -334,6 +342,7 @@ impl Frame {
 
     /// Execute a single instruction.
     #[allow(clippy::cognitive_complexity)]
+    #[cfg_attr(feature = "flame-it", flame("Frame"))]
```
Comment (windelbouwman) [inline vm/src/frame.rs]: Do we need to explicitly enable all trace points?

**Comment by windelbouwman** [comment]
Off topic: why does the importlib loading take so long? Is it a whole lot of code?

**Comment by coolreader18** [comment]
re: importlib: The main problem is that it is a fair bit of code, and RustPython isn't all that optimized. See #1047.

re: automatically tagging all fns: see llogiq/flamer#45. I tried to do that initially, and there were a lot of issues with it. Even so, manually tagging all fns in `frame.rs` with `#[flame]` gave way too much irrelevant data, and the file size was huge as well (36MB compared to the 1MB with the current version here). e.g., do we really need to see how long each individual exception creation function takes?

**Comment by coolreader18** [comment]
I had to use a modified local clone of flame to get [this flamegraph](https://gistcdn.githack.com/coolreader18/4b173399907966958f0ea3244ca07593/raw/323cb62d713bf53b7ea5171957ff0c8a9fbdc411/flame-graph.html) due to a bug (TyOverby/flame#47), but something interesting that you can see is that in `_install_external_importers`, when it imports `_frozen_importlib_external`, a fairly large portion of the time spent in `__import__` is before it even starts executing the module (`Frame::run(<module>)`). Maybe we can detect when it's importing `_frozen_importlib_external` from `_frozen_importlib` and use Rust to import it?

**Comment by coolreader18** [comment]
Here it is with a simple optimization to skip `__import__` in that case: [link](https://gistcdn.githack.com/coolreader18/4b173399907966958f0ea3244ca07593/raw/e2c491c73ded65916544cd73e1ef9f387062d572/flame-graph-optimized.html). Shaved off 10ms, not bad!

**Comment by palaviv** [comment]
CPython actually have large parts of the `__import__` function implemented in C. We can do the same in Rust for optimization.

**Pull request merged**